### PR TITLE
Ensure consistency of system flag on IndexMetadata after diff is applied

### PR DESCRIPTION
**Pull request by cwperks**
### Description

This PR ensures the consistency of the `isSystem` flag between leader and follower after the IndexMetadataDiff is applied when computing new cluster state. 

This inconsistency can appear during rolling upgrades if nodes of a new version know of an index as a system index, but nodes of a previous version treat the same index as a regular index. For example, if a plugin retroactively declares an index to be a system index through SystemIndexPlugin.getSystemIndexDescriptors (Example [PR](https://github.com/opensearch-project/observability/pull/1837)) there can be a discrepancy in the IndexMetadata (for the same version) between the cluster manager node and other nodes in the cluster.

The reason this happens is because when the IndexMetadataDiff is applied, its taking the value from the previous metadata on the node instead of the new metadata that it has received from the incoming publishRequest. As a result, the diff [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/coordination/PublicationTransportHandler.java#L207-L210) is being computed incorrectly. It should take the value from the diff instead of from the previous index metadata.

### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/16643

### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by Bukhtawar — APPROVED** [review]
Thanks for the quick fix. Ship it!

**Review by shwetathareja — APPROVED** [review]

**Comment by owaiskazi19** [comment]
@cwperks this change needs a changelog entry

**Comment by cwperks** [comment]
> @cwperks this change needs a changelog entry

Added

**Comment by cwperks** [comment]
Resolved conflicts in CHANGELOG

**Comment by cwperks** [comment]
Resolved conflicts on CHANGELOG again. Can a maintainer help to merge this PR?

File: server/src/test/java/org/opensearch/cluster/metadata/IndexMetadataTests.java [hunk 1]
```diff
@@ -88,6 +92,36 @@ protected NamedXContentRegistry xContentRegistry() {
         return new NamedXContentRegistry(IndicesModule.getNamedXContents());
     }
 
+    // Create the index metadata for a given index, with the specified version.
+    private static IndexMetadata createIndexMetadata(final Index index, final long version) {
```
Comment (sandeshkr419) [inline server/src/test/java/org/opensearch/cluster/metadata/IndexMetadataTests.java:96]: nitpick on code refactoring:

```
private static IndexMetadata createIndexMetadata(final Index index, final long version) {
    return createIndexMetadata(index, version, false (or the default value));
}
```

unless setting to default vs not setting the `.system()` achieves a different functionality?

Comment (cwperks) [reply server/src/test/java/org/opensearch/cluster/metadata/IndexMetadataTests.java:96]: Pushed this change to reuse the method with default.

**Review by sandeshkr419 — APPROVED** [review]

**Comment by cwperks** [comment]
Resolved conflicts in CHANGELOG

**Review by Bukhtawar — APPROVED** [review]

**Comment by cwperks** [comment]
Creating a manual backport now

**Pull request merged**